### PR TITLE
feat: support explicitly setting agency and reporting period in the ARPA Reporter upload api (#895)

### PR DESCRIPTION
**Pull request by NatHillardUSDR**
Backend change to allow setting agency and reporting period in uploads. Not yet populated on the frontend.

DRAFT until I can verify a few scenarios / have a few examples to work against

## Description
Adding a simple change to apply the logic from the original ticket when uploading values for agencyId and reportingPeriodId

## Screenshots / Demo Video
N/A, backend only change

## Testing
* Not yet possible to test this, as values are not being populated in the UI, but you should be able to upload workbooks as usual without side effects

### Automated and Unit Tests
- [x] Added Unit tests

### Manual tests for Reviewer
- [x] Added steps to test feature/functionality manually

## Checklist
- [x] Provided ticket and description
- [x] Provided screenshots/demo
- [x] Provided testing information
- [x] Provided adequate test coverage for all new code
- [x] Added PR reviewers

### CODE REVIEW

**Comment by NatHillardUSDR** [comment]
Fixed the logic per convo with Truman in #895 - should be ready to go!

File: packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js [hunk 1]
```diff
@@ -34,36 +62,54 @@ describe('persist-upload', () => {
...
     describe('persistUploadToFS', () => {
-        const persistUploadToFS = persistUpload.__get__('persistUploadToFS');
+        const persistUploadToFS = persistUploadModule.__get__('persistUploadToFS');
         let fsMock;
         const filename = 'test.xlsx';
         const upload = { filename, id: 1 };
         const filePath = uploadFSName(upload);
         const buffer = Buffer.from('test data');
 
+        beforeEach(() => {
```
Comment (trumanc) [inline packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js:91]: Is this duplicate beforeEach call (and the duplicate afterEach below) intentional? It looks like maybe a merge error somewhere resulted in two copies of both

File: packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js [hunk 2]
```diff
@@ -86,37 +132,128 @@ describe('persist-upload', () => {
...
+        getReportingPeriod.resolves(reportingPeriod);
+        persistUploadModule.__set__('getReportingPeriod', getReportingPeriod);
+
+        const result = await getValidReportingPeriodId(reportingPeriod.id);
+
+        expect(getReportingPeriod.calledOnceWith(reportingPeriod.id)).to.be.true;
+        expect(result).to.equal(reportingPeriod.id);
+    });
+
+    it('should throw a ValidationError when the reporting period does not exist', async () => {
```
Comment (trumanc) [inline packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js:234]: In addition to checking the "does not exist" case, a more critical safety check might be the "reporting period does exist for a different tenant" case

**Review by trumanc — COMMENTED** [review]
Overall this looks great! Left a few comments around making sure error messages are appropriate for their audiences, and some nits around variable names and one potentially important new test case.

File: packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js (hunk 1 above)
Comment (NatHillardUSDR) [reply packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js]: Oof, good catch, got this when resolving merge conflicts.

File: packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js (hunk 2 above)
Comment (NatHillardUSDR) [reply packages/server/__tests__/arpa_reporter/server/services/persist-upload.spec.js:229]: Interesting! I was initially thinking this was covered by the `reporting-period.spec.js` test, and thus a few layers deeper than the "unit" of this method, but then I realized, `getReportingPeriod` actually returns `undefined` if the conditions aren't met, so we can represent this with the mocks - I added a test to represent this

File: packages/server/src/arpa_reporter/services/persist-upload.js [hunk 3]
```diff
@@ -55,19 +91,93 @@ async function persistUploadToFS(upload, buffer) {
...
+    // itself in validate-upload. We leave it as null here.
+    if (!agencyId) {
+        return null;
+    }
+    // Otherwise, we need to make sure the user is associated with the agency
+    const userRecord = await getUser(userId);
+    const tenantAgencies = await getTenantAgencies(userRecord.tenant_id);
+    const agency = tenantAgencies.find((a) => a.id === Number(agencyId));
+    if (!agency) {
+        throw new ValidationError(`Supplied agency ID ${agencyId} does not correspond to an agency in the user's tenant ${userRecord.tenant_id}`);
```
[the remark this answers is not in the document]
Comment (NatHillardUSDR) [reply packages/server/src/arpa_reporter/services/persist-upload.js]: Really appreciate you catching / thinking about this - fixed to add that message.

File: packages/server/src/arpa_reporter/services/persist-upload.js [hunk 4]
```diff
@@ -55,19 +91,93 @@ async function persistUploadToFS(upload, buffer) {
...
+    return agencyId;
+}
+
+/**
+ * Validate the reporting period ID
+ * @param {string} reportingPeriodId
+ * @returns {string}
+ * @throws {ValidationError}
+ */
+async function getValidReportingPeriodId(reportingPeriodId) {
```
[the remark this answers is not in the document]
Comment (NatHillardUSDR) [reply packages/server/src/arpa_reporter/services/persist-upload.js]: Yeah I went back and forth on this name a few times 😭 -- I like `ensureValidReportingPeriodId` -- I've updated the other method names to match as well.

**Review by NatHillardUSDR — COMMENTED** [review]
Thank you for your review and comments! Great catches, and definitely help with clarity to both volunteers and end users. I believe I've addressed them all, let me know if additional changes would be helpful.

**Review by trumanc — APPROVED** [review]

**Pull request merged**